### PR TITLE
MP-5715: Marketplace Product Offer Prices

### PR DESCRIPTION
**Pull request by zssamoylov**
- Ticket: https://spryker.atlassian.net/browse/MP-5715

| MODULE     | DESCRIPTION                |
|------------|----------------------------|
| Price | Handles product pricing and provides plugins for products to populate prices   |
| PriceProduct | Provides product price related functionality, price persistence, current price resolvers per currency/price mode    |
| PriceProductDataImport | Imports demo data for product prices    |
| PriceProductOffer | Provides product offer price related functionality, price persistence, current price resolvers per currency/price mode    |
| PriceProductOfferDataImport | Imports demo data for product offer prices    |
| PriceProductOfferGui | Backoffice UI Interface for managing prices for product offers    |
| PriceProductOfferStorage | Provides functionality to store data about product offer prices in the storage   |
| PriceProductOfferVolume | Provides functionality to handle volume prices for product offers    |
| PriceProductOfferVolumeGui | Backoffice UI Interface for managing volume prices for product offers    |
| PriceProductStorage | Provides functionality to store data about product prices in the storage    |
| PriceProductVolume | Provides functionality to handle volume prices for products   |
| PriceProductVolumeWidget | Provides widget to show information about volume price for product   |
| PriceWidget | Provides functionality for price mode switcher    |
| PriceProductOfferExtension | Provides plugin interfaces for extending PriceProductOffer module functionality    |
| PriceProductOfferStorageExtension | Provides plugin interfaces used by Price Product Offer Storage bundle    |
| PriceProductOfferVolumesRestApi | Provides plugin(s) to add product-offer-volume-prices to the product-offer-prices.   |
| ProductOfferPricesRestApi | Provides Rest API endpoints to manage product offer prices   |
| ProductOfferPricesRestApiExtension | Provides plugin interfaces for extending the ProductOfferPricesRestApi module    |

### CODE REVIEW

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 1]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
 {% endinfo_block %}
 
-## Related Developer articles
+## Marketplace Product Offer Prices - Overview
+Product offer prices can be defined for each product offer by merchants. 
+If price has not defined for product offer, will be used price from concrete product which is tied to the product offer by default.
+Each price should have type. 
+The price type it is an entity to define what type of prices will be used for selling. 
+There are 2 price types defined in the Spryker system, it's ORIGINAL and DEFAULT. 
+You can add own price types and use it in your app.
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:22]: Would be nice to describe how to add ad new price type (according to this document https://spryker.atlassian.net/wiki/spaces/DOCS/pages/2733801742/WIP+Marketplace+Back+End#Marketplace-Product it should be described)

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 2]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
 
 {% endinfo_block %}
 
-## Related Developer articles
+## Marketplace Product Offer Prices - Overview
+Product offer prices can be defined for each product offer by merchants. 
+If price has not defined for product offer, will be used price from concrete product which is tied to the product offer by default.
+Each price should have type. 
+The price type it is an entity to define what type of prices will be used for selling. 
+There are 2 price types defined in the Spryker system, it's ORIGINAL and DEFAULT.
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:21]: https://documentation.spryker.com/docs/prices-overview Can be a text from this document be reused? Looks clearer

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 3]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+Product offer prices can be defined for each product offer by merchants. 
+If price has not defined for product offer, will be used price from concrete product which is tied to the product offer by default.
+Each price should have type. 
+The price type it is an entity to define what type of prices will be used for selling. 
+There are 2 price types defined in the Spryker system, it's ORIGINAL and DEFAULT. 
+You can add own price types and use it in your app.
+The price can have gross or net value which can be used based on a price mode selected by customer in Storefront. 
+You can have shop running in both modes and select net mode for business customer, for example. 
+Price also has currency and store assigned to it.
+To support product offer prices has been added *PriceProductOffer* table to connect *PriceProductStore* and *ProductOffer* tables.
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:26]:
```suggestion
To support product offer prices  *PriceProductOffer* DB table has been added to connect *PriceProductStore* and *ProductOffer* tables.
```

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 4]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+To store data about product offer prices has been added *ProductConcreteProductOfferPriceStorage*.
+The data updated, when creating/updating/deleting/activating/deactivating concrete products and when create price product offers. 
+Also for product offers has been added support volume prices.
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:29]:
```suggestion
Also product offers support volume prices. 
```

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 5]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+You can add own price types and use it in your app.
+The price can have gross or net value which can be used based on a price mode selected by customer in Storefront. 
+You can have shop running in both modes and select net mode for business customer, for example. 
+Price also has currency and store assigned to it.
+To support product offer prices has been added *PriceProductOffer* table to connect *PriceProductStore* and *ProductOffer* tables.
+To store data about product offer prices has been added *ProductConcreteProductOfferPriceStorage*.
+The data updated, when creating/updating/deleting/activating/deactivating concrete products and when create price product offers. 
+Also for product offers has been added support volume prices. 
+So merchants can add volume prices for product offers and based on chosen quantity of product offers customer will see the price on the Storefront.
+Volume prices for product offers work the same as for products.
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:31]:
```suggestion
So merchants can add volume prices for product offers and based on chosen quantity of items a customer will see the corresponding price on the Storefront.
Volume prices for product offers work the same as for products.
```

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 6]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+| PriceProductOffer | Provides product offer price related functionality, price persistence, current price resolvers per currency/price mode    |
+| PriceProductOfferDataImport | Imports demo data for product offer prices    |
+| PriceProductOfferGui | Backoffice UI Interface for managing prices for product offers    |
+| PriceProductOfferStorage | Provides functionality to store data about product offer prices in the storage   |
+| PriceProductOfferVolume | Provides functionality to handle volume prices for product offers    |
+| PriceProductOfferVolumeGui | Backoffice UI Interface for managing volume prices for product offers    |
+| PriceProductStorage | Provides functionality to store data about product prices in the storage    |
+| PriceProductVolume | Provides functionality to handle volume prices for products   |
+| PriceProductVolumeWidget | Provides widget to show information about volume price for product   |
+| PriceWidget | Provides functionality for price mode switcher    |
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:52]: Please add optional modules

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 7]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+
+## Domain model
+![Entity diagram](https://confluence-connect.gliffy.net/embed/image/0ad490bb-f21f-4e4a-b6eb-e0102a8c7b42.png?utm_medium=live&utm_source=confluence)
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:55]: Will it look better if to bread `ProductConcreteProductOfferPriceStorage` to 2 lines?

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 8]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+To store data about product offer prices has been added *ProductConcreteProductOfferPriceStorage*.
+The data updated, when creating/updating/deleting/activating/deactivating concrete products and when create price product offers. 
+Also for product offers has been added support volume prices. 
+So merchants can add volume prices for product offers and based on chosen quantity of product offers customer will see the price on the Storefront.
+Volume prices for product offers work the same as for products.
+
+To learn more about prices and volume prices, see: [Prices](https://documentation.spryker.com/docs/prices-overview), [Volume Prices](https://documentation.spryker.com/docs/volume-prices-overview)
+
+## Module dependency graph
+![Entity diagram](https://confluence-connect.gliffy.net/embed/image/f128877d-eb61-4d87-b1af-5f166eb45c45.png?utm_medium=live&utm_source=confluence)
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:36]: - please add optional modules
- PriceProductOfferStorage -> ProductOffer, PriceProductOfferDataImport -> ProductOffer, PriceProductOffer -> -> ProductOffer, 
- I thins it is important to show that PriceProduct depends on Product, what do you think? (in this case dependencies from other module to Product should be added)

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md [hunk 9]
```diff
@@ -13,8 +13,48 @@ To learn more about the feature and to find out how end users use it, see [Produ
...
+If price has not defined for product offer, will be used price from concrete product which is tied to the product offer by default.
+Each price should have type. 
+The price type it is an entity to define what type of prices will be used for selling. 
+There are 2 price types defined in the Spryker system, it's ORIGINAL and DEFAULT. 
+You can add own price types and use it in your app.
+The price can have gross or net value which can be used based on a price mode selected by customer in Storefront. 
+You can have shop running in both modes and select net mode for business customer, for example. 
+Price also has currency and store assigned to it.
+To support product offer prices has been added *PriceProductOffer* table to connect *PriceProductStore* and *ProductOffer* tables.
+To store data about product offer prices has been added *ProductConcreteProductOfferPriceStorage*.
```
Comment (ilyakubanov) [inline docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:27]: Sounds like just for saving this data, not clear why it is needed

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md (hunk 2 above)
Comment (zssamoylov) [reply docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:21]: I've added link to this document, below.

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md (hunk 7 above)
Comment (zssamoylov) [reply docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:61]: updated.

File: docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md (hunk 8 above)
Comment (zssamoylov) [reply docs/marketplace/dev/feature-walkthroughs/202108.0/marketplace-product-offer-prices-feature-walkthrough.md:42]: I'm not sure that we need this information.
We provided links to related articles, including product offers.

**Review by gibmirmilch — APPROVED** [review]

**Pull request merged**